rfcs/0005-pablo-distribution-assets/0005-pablo-distribution.adoc

```diff
@@ -0,0 +1,632 @@
+        [RewardDisbursementHook]
+        [RewardsConfig] --> [RewardAccumulationHook]
+    }
+}
+
+[fNFT Storage/Logic] <<exists>>
+[RewardDisbursementHook] <<remove>>
+
+node "Pallet-Pablo" {
+    [Pool] --> StakingReward : transfer\ntrading fees\nfor PBLO stakers
+    [Pool] --> Staking : stake LP tokens
+    [Pool] --> [Pool] : accumulate LP\ntrading fees
+    [FeeConfig] --> [Pool]
+}
+
+[Pool] <<exists>>
+
+
+node "Governance" {
+    [Configuration] --> [FeeConfig]
+    [Treasury]--> StakingReward: Transfer LP, PBLO\nor other token rewards
+}
+----
+
+TODO: What to do for part of protocol fees that should be transferred to treasury eventually as treasury does not stake it's PBLO?
+
+=== Pallet-Pablo
+
+In order to 1. support LP staking 2. LP trading fee distribution and 3. PBLO staking reward using trading fees, following changes are proposed for https://github.com/ComposableFi/composable/tree/main/frame/pablo[Pallet-Pablo].
+
+[#_feeconfig]
+==== FeeConfig
+
+Each pool in Pablo defines a fee percentage to be charged for each trade.Except for LBPs other pools also define an owner fee that is a percentage out of the main trading fee. The `FeeConfig` is a new abstraction over all fees that could be charged on a pool to allow for extension. At this time a 100% of the owner fee should be defined as a new field `protocol_fee`.
+
+[plantuml,images/pablo-fee-config,png]
+----
+class FeeConfig {
+    +fee_rate: Permill
+    +owner_fee_rate: Permill,
+    +protocol_fee_rate: Permill,
+}
+----
+
+*Existing code must be modified to use this data structure*.
+
+Given this,
+----
+fee = // calculation depends on the pool type: based on the fee_rate
+owner_fee = fee * owner_fee_rate * (1 - protocol_fee_rate);
+protocol_fee = owner_fee * protocol_fee_rate;
+----
+For all pools launched at the Picasso launch following values would be set for these configs
+----
+owner_fee_rate = 20%
+protocol_fee_rate = 100% // all owner fees goes to composable to be distributed as rewards
+----
+
+==== LP Trading Fee Distribution
+
+LPs trading fees are calculated and kept as part of the pool liquidity in Pablo. When LPs remove liquidity from the pool the trading fees are automatically redeemed according their pool LP ratio, check https://hackmd.io/@HaydenAdams/HJ9jLsfTz#Fee-Structure[reference]. This results in trading fee share being diluted overtime for smaller pools as follows.
+
+After stem:[n] trades and stem:[m] liquidity additions,
+
+trading fees stem:[= sum a_n]
+
+total liquidity stem:[= sum l_m]
+
+fees and liquidity returned for an LP amount stem:[l = l / (sum l_m) * (sum l_m + sum a_n)]
+
+stem:[= l + l / (sum l_m) * sum a_n]
+
+trading fees received stem:[= x = l / (sum l_m) * sum a_n]
+
+"When pool size stem:[sum l_m] increases the amount of trading fees received stem:[x] reduces for a particular LP position.
+
+For large pool sizes of stem:[sum l_m] (steady state) this effect is negligible, hence it's a good enough strategy to distribute fees.
```

Comment:
not sure what is comment about. this works for anu pool sizes and by design honest (if we keep in mind using dillution protection)